Automattic/wp-calypso
# Checkout: automate payment processor by standardizing responses

#### Changes proposed in this Pull Request

This is a change to the `composite-checkout` package that automates every payment processor so that they are no longer responsible for setting the transaction status manually. They can each now just call the `onClick` handler (passed as a prop to the button), and the payment processor function will take care of the work of converting the submitted data into a standardized response. That response will then trigger transaction status changes automatically.

This requires the payment processor functions to take on any additional processing required to transform the actual server response into one of a set of standard responses that the `composite-checkout` package can handle. Currently there are three such responses represented by the returned Promise resolving with one of three codes: SUCCESS, REDIRECT, and NOOP. There's also the failure state which is represented by the Promise rejecting. SUCCESS and REDIRECT were already mostly handled by the transaction status system but had to be set manually; now the response will do that. The NOOP response allows the payment method to take its own actions if it wishes (eg: the WeChat payment method sometimes displays a QR code as a final state and handles its own redirect and error states).

Example of a payment method button updated to use the new system where you can see the removed boilerplate:

<img width="1005" alt="Screen Shot 2020-10-10 at 12 39 37 AM" src="https://user-images.githubusercontent.com/2036909/95645877-31626600-0a91-11eb-97ad-c673581922e4.png">

This greatly simplifies making a payment method, although the amount of processing the server response is the same; the work is just moved to the payment processor function with the transaction status boilerplate removed.

In terms of the code, this also will make it much easier in the future to convert the payment method objects to TypeScript as it removes their most complex part. It also allows us to standardize on several patterns around redirects and error messages. Altogether these changes should lead to more reliable and flexible code with fewer bugs.

This PR only changes two payment methods: credit card (which actually includes the built-in credit card method used only for the demo, the stripe credit card method, and the ebanx method), and existing cards. Other payment methods will be converted in future PRs:

- PayPal: https://github.com/Automattic/wp-calypso/pull/46431
- Free and Full Credits: https://github.com/Automattic/wp-calypso/pull/46430
- Apple Pay: https://github.com/Automattic/wp-calypso/pull/46429
- Stripe Redirect methods: https://github.com/Automattic/wp-calypso/pull/46428

#### Testing instructions

- Begin a payment using PayPal and verify that once you click the submit button, you are redirected to PayPal (it should not be changed by this PR so no need to test the whole purchase).
- Complete a payment using a credit card processed through Stripe.
  - Use a [stripe test card](https://stripe.com/docs/testing).
  - Verify that the payment succeeds and that you are redirected to the thank-you page.
- Complete a payment using a credit card processed through Stripe that requires 3DS.
  - Use a [stripe test card](https://stripe.com/docs/testing) that requires 3D Secure 1 or 2.
  - Verify that the 3DS modal is displayed, that the payment succeeds, and that you are redirected to the thank-you page.
- Complete a payment using an existing credit card.
  - Verify that the payment succeeds and that you are redirected to the thank-you page.
- Complete a payment using an existing credit card that requires 3DS.
  - Verify that the 3DS modal is displayed, that the payment succeeds, and that you are redirected to the thank-you page.
- Complete a payment using a credit card processed through Ebanx.
  - Apply D48147-code to your sandbox to pretend to be in Brazil and sandbox the API.
  - Set your currency to `BRL`.
  - In the contact information step, choose "Brazil" for the country.
  - In the final step, choose "Credit or debit card".
  - Verify that the form asks for all the Ebanx fields in addition to the credit card fields (Taxpayer Identification Number, Phone, etc.)
  - Use an [ebanx test card](https://developer.ebanx.com/docs/resources/testCreditCards/) and [test customer info](https://developer.ebanx.com/docs/resources/testCustomerData/).
  - Verify that the payment succeeds and that you are redirected to the thank-you page.
